File: simulator/spells/cloud_of_daggers.py
from functools import cache

from cachetools import cached
from cachetools.keys import hashkey

from ..actions.action_types import BonusAction
from ..battle_map import Map, map_position_toggled_cache, map_toggled_cache_with_key
from ..combatant_coords import Coords
from ..effects.aoe_square_effect import AoeSquareEffect
from ..effects.effect import EffectType
from ..effects.limited_duration_effect import LimitedDurationEffect
from ..spells.spell import SpellStats
from ..misc import DamageType, avg_roll, roll_spell_dmg, Conditions
from ..actions.actoid import Actoid, ActoidFlags
from ..threat_interfaces import DirectThreat, DirectThreatFactory, AoEThreat
import numpy as np

class CloudOfDaggersFactory(DirectThreatFactory):
    level = 2
    range = SpellStats.Range.FEET_60.value
    target = SpellStats.Target.BOX_5
    duration = SpellStats.Duration.INSTANTANEOUS
    concentration = True
    type = SpellStats.Type.HARMFUL
    dmg_type = DamageType.Slashing

    def __init__(self, action_type, caster, **kwargs):
        super().__init__()
        self.action_type = action_type  # SPIKE_GROWTH, QUICKENED_SPIKE_GROWTH
        self.dmg_dice = "4d4"
        self.combatant = caster


    def __str__(self):
        """
        Important for FSM building
        """
        return "CloudOfDaggersFactory"


    def get_ability_name(self):
        return "Cloud of Daggers"


    def find_best_args(self, combatant):
        # TODO maybe find a smarter placement for this
        coord, _, _ = Map.get().find_best_placement_harmful_square(self.combatant, CloudOfDaggersFactory.range, 1)
        return coord

    def create_all(self):
        # Here there really is no need to iterate over all coords. Just find the best score
        return [CloudOfDaggers(self.find_best_args(self.combatant), self)]

    def create(self, coord):
        return CloudOfDaggers(coord, self)

    def calculate_threat_to_target(self, target, **kwargs):
        """
        Calculates threat to one specific target
        """
        return avg_roll(self.dmg_dice)

    def calculate_threat_to_target_delta(self, target, modifiers, *args, **kwargs):
        """
        Calculates the threat delta of the factory to a specific target given stat modifications
        """
        return 0  # No need

    def calculate_max_threat(self):
        return CloudOfDaggers(self.find_best_args(self.combatant), self).calculate_threat()

class CloudOfDaggers(Actoid, LimitedDurationEffect, AoeSquareEffect, DirectThreat, AoEThreat):

    def __init__(self, coord, factory,  **kwargs):
        super().__init__(actoid_flags=ActoidFlags.IS_SPELL)
        LimitedDurationEffect.__init__(self, factory.combatant, turns=10)
        AoeSquareEffect.__init__(self, factory.combatant, coord, SpellStats.TRANSLATE_BOX[CloudOfDaggersFactory.target])
        self.factory = factory

    def __str__(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_CLOUD_OF_DAGGERS else "") + f"Cloud of Daggers at {np.squeeze(self.origin)}"

    def get_effect_type(self):
        return EffectType.CLOUD_OF_DAGGERS

    def shorthand_str(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_CLOUD_OF_DAGGERS else "") + f"Cloud of Daggers"

    def on_start_of_turn(self, combatant):
        dmg = roll_spell_dmg(self.factory.dmg_dice)
        combatant.receive_dmg(dmg, CloudOfDaggersFactory.dmg_type)

    def on_end_of_turn(self, combatant):
        pass

    def on_enter(self, combatant):
        dmg = roll_spell_dmg(self.factory.dmg_dice)
        combatant.receive_dmg(dmg, CloudOfDaggersFactory.dmg_type)

    def on_move_within(self, combatant):
        return 0

    def on_exit(self, combatant):
        pass

    def is_affecting(self, combatant):
        coords = self.get_affected_coords()
        battle_map = Map.get()
        return battle_map.get_hop_distance_coords(battle_map.get_combatant_position(combatant).get(), coords) == 0

    def activate(self):
        Map.get().effect_tracker.add(self)
        self.factory.combatant.concentration_effect = self

    def deactivate(self):
        self.factory.combatant.break_concentration()

    @map_position_toggled_cache
    def calculate_threat(self, **kwargs):
        battle_map = Map.get()
        affected = battle_map.get_combatants_affected_by_aoe(self.factory.combatant, CloudOfDaggersFactory.target, CloudOfDaggersFactory.type, self.origin)
        acc = 0
        for aff in affected:
            acc += (1 if battle_map.teams.are_enemies(self.factory.combatant, aff) else -3) * avg_roll(self.factory.dmg_dice)
        return acc

    def clear_cache(self):
        self.calculate_threat.cache_clear()
        #self.get_eligible_coords.cache_clear()

    def calculate_threat_delta(self, modifiers, *args, **kwargs):
        return 0  # Not relevant for this ability

    def threat_on_end_of_turn(self, target, *args, **kwargs):
        return 0

    def threat_on_enter(self, target, *args, **kwargs):
        return avg_roll(self.factory.dmg_dice)

    def threat_on_start_of_turn(self, target, *args, **kwargs):
        return avg_roll(self.factory.dmg_dice)

    def threat_on_move_within(self, target, *args, **kwargs):
        return 0

    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        if self.factory.combatant.get_swallower():
            return None
        if not self.factory.combatant.is_affected_by_any(Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
            return battle_map.get_free_coords_in_cartesian_range(Coords(self.origin),  # not actually combatant coords
                                                                 distances,
                                                                 inflate_to_size=self.factory.combatant.size,
                                                                 rng=CloudOfDaggersFactory.range, combatant=self.factory.combatant)
        elif battle_map.get_cartesian_distance_coords(battle_map.get_combatant_position(self.factory.combatant).get(), np.array([self.origin])) <= CloudOfDaggersFactory.range:
            return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]
        return None

File: c++/spells/cloud_of_daggers.cpp
#include "spells/cloud_of_daggers.hpp"
#include "core/battle_map.hpp"
#include "core/misc.hpp"
#include "core/teams.hpp"
#include "core/conditions.hpp"
#include "core/geometry.hpp"
#include <memory>

namespace enc
{

  CloudOfDaggersFactory::CloudOfDaggersFactory(AbilityType abilityType, Combatant *caster, Resource *resource)
      : DirectThreatFactory("CloudOfDaggersFactory", "Cloud of Daggers", caster, abilityType), _abilityType(abilityType), _resource(resource),
        _dmgDice({4, 4})
  {}

  std::vector<std::shared_ptr<Actoid>> CloudOfDaggersFactory::createAll(void *previousActionInDag)
  {
    BattleMap &battleMap = BattleMap::getInstance();
    auto [coord, maxScore, affectedCombatants]
      = battleMap.findBestPlacementHarmfulSquare(_combatant, static_cast<int>(CloudOfDaggersFactory::range), 1);
    return {std::make_shared<CloudOfDaggers>(coord, *this)};
  }

  std::shared_ptr<Actoid> CloudOfDaggersFactory::create(void *target)
  {
    Coord *coord = static_cast<Coord *>(target);
    return std::make_shared<CloudOfDaggers>(*coord, *this);
  }

  double CloudOfDaggersFactory::calculateThreatToTarget(Combatant *target, const Kwargs &kwargs) const { return avgRoll(_dmgDice); }

  double CloudOfDaggersFactory::calculateThreatToTargetDelta(Combatant *target, const ThreatModifiers &modifiers) const
  {
    return 0.0; // No need
  }

  double CloudOfDaggersFactory::calculateMaxThreat() const
  {
    BattleMap &battleMap = BattleMap::getInstance();
    auto [coord, maxScore, affectedCombatants]
      = battleMap.findBestPlacementHarmfulSquare(_combatant, static_cast<int>(CloudOfDaggersFactory::range), 1);
    return CloudOfDaggers(coord, *this).calculateThreat(Kwargs());
  }

  std::string CloudOfDaggers::toString() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_CLOUD_OF_DAGGERS) ? "Quickened " : "";
    return prefix + "Cloud of Daggers at (" + std::to_string(_coord[0]) + ", " + std::to_string(_coord[1]) + ")";
  }

  std::string CloudOfDaggers::shorthandStr() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_CLOUD_OF_DAGGERS) ? "Quickened " : "";
    return prefix + "Cloud of Daggers";
  }

  double CloudOfDaggers::calculateThreat(const Kwargs &kwargs)
  {
    BattleMap &battleMap = BattleMap::getInstance();
    Teams &teams = Teams::getInstance();

    std::vector<Combatant *> affectedCombatants = battleMap.getCombatantsAffectedByBoxAoE(CloudOfDaggersFactory::target, _coord);

    double acc = 0.0;
    for(auto *aff : affectedCombatants)
      {
        double avgDmg = avgRoll(_factory._dmgDice);
        acc += (teams.areEnemies(*_factory._combatant, *aff) ? 1.0 : -3.0) * avgDmg;
      }
    return acc;
  }

  double CloudOfDaggers::calculateThreatDelta(const ThreatModifiers &modifiers) const
  {
    return 0.0; // Not relevant for this ability
  }

  std::optional<CoordVector>
  CloudOfDaggers::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    Combatant *swallower = _factory._combatant->getSwallower();
    if(swallower)
      {
        return std::nullopt;
      }

    BattleMap &battleMap = BattleMap::getInstance();
    if(!_factory._combatant->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        return battleMap.getFreeCoordsInCartesianRange(Coords(_coord), distances, _factory._combatant->getSize(),
                                                       static_cast<int>(CloudOfDaggersFactory::range), _factory._combatant->_instanceId);
      }
    else if(getCartesianDistanceCoords(battleMap.getCombatantCoordinates(*_factory._combatant), Coords(_coord))
            <= static_cast<int>(CloudOfDaggersFactory::range))
      {
        return CoordVector{battleMap.getCombatantCoordinates(*_factory._combatant).getRoot()};
      }

    return std::nullopt;
  }

  void CloudOfDaggers::activate(const Kwargs &kwargs)
  {
    _factory._combatant->setConcentrationEffect(Effect::shared_from_this());
  }

  void CloudOfDaggers::deactivate() { _factory._combatant->breakConcentration(); }

  bool CloudOfDaggers::deactivateForCombatant(Combatant *combatant)
  {
    assert(false);
    return false;
  }

  void CloudOfDaggers::onEnter(Combatant *combatant)
  {
    int damage = rollDice(_factory._dmgDice);
    combatant->receiveDmg(damage, CloudOfDaggersFactory::dmgType);
    BattleMap::getInstance().removeCombatantIfDead(*combatant);
  }

  void CloudOfDaggers::onMoveWithin(Combatant * /*combatant*/) {}

  void CloudOfDaggers::onExit(Combatant * /*combatant*/) {}

  void CloudOfDaggers::onStartOfTurn(Combatant *combatant)
  {
    int damage = rollDice(_factory._dmgDice);
    combatant->receiveDmg(damage, CloudOfDaggersFactory::dmgType);
    BattleMap::getInstance().removeCombatantIfDead(*combatant);
  }

  void CloudOfDaggers::onEndOfTurn(Combatant * /*combatant*/) {}

  double CloudOfDaggers::threatOnEnter(Combatant * /*target*/, const Kwargs & /*kwargs*/) const
  {
    return avgRoll(_factory._dmgDice);
  }

  double CloudOfDaggers::threatOnStartOfTurn(Combatant * /*target*/, const Kwargs & /*kwargs*/) const
  {
    return avgRoll(_factory._dmgDice);
  }

  double CloudOfDaggers::threatOnMoveWithin(Combatant * /*target*/, const Kwargs & /*kwargs*/) const { return 0.0; }

  double CloudOfDaggers::threatOnEndOfTurn(Combatant * /*target*/, const Kwargs & /*kwargs*/) const { return 0.0; }
} // namespace enc
